Use jstree

alimanfoo opened the pull request:
This PR modifies the tree HTML representation to use jstree. Also adds a "level" parameter to tree() to allow limiting depth to which tree is printed.

alimanfoo commented:
[example notebook](http://nbviewer.jupyter.org/github/alimanfoo/zarr/blob/5cc474a3ff21afc7b0173aacb13aa7aae1b4c6c8/notebooks/repr_tree.ipynb) is working in nbviewer

alimanfoo commented:
cc @jakirkham

jakirkham commented on notebooks/repr_tree.ipynb:
```diff
@@ -6,13 +6,9 @@
-    "except NameError:\n",
-    "    unicode = str"
+    "import sys\n",
+    "sys.path.insert(0, '..')\n",
```
Why is this needed?

alimanfoo replied:
So zarr is on the path. How did you do that, did you set PYTHONPATH?

jakirkham commented on zarr/util.py:
```diff
@@ -312,117 +313,126 @@ def _repr_html_(self):
-    result += custom_html_sublist(group, indent=indent)
-    result += "</ul>\n"
-    result += "</div>\n"
+def tree_html(group, expand, level):
```
This is probably my error in the first place, but should we just make this a method of `TreeViewer`?

alimanfoo replied:
Yes, could do.

jakirkham commented on zarr/util.py:
```diff
@@ -312,117 +313,126 @@ def _repr_html_(self):
+tree_array_icon = 'fa fa-table'
+# alternatives...
+# tree_group_icon: 'jstree-folder'
+# tree_array_icon: 'jstree-file'
```
Maybe these should be variables in `TreeViewer`.

alimanfoo replied:
Yes could be. I was trying to think of best way to make it so a really enthusiastic user could change the default icons if they want to use their favourite emojis instead :-)

jakirkham replied:
Was thinking about having static variables of the class as an option as well. Maybe that would work better?

alimanfoo commented on zarr/util.py:
```diff
@@ -312,117 +313,126 @@ def _repr_html_(self):
+
+    # include CSS for jstree default theme
+    css_url = '//cdnjs.cloudflare.com/ajax/libs/jstree/3.3.3/themes/default/style.min.css'
+    result += '<link rel="stylesheet" href="{}"/>'.format(css_url)
```
I think that just gets handled automatically. I've just been copying examples there.

jakirkham replied:
Guess so. Interesting. Also this [SO question]( https://stackoverflow.com/q/9374203 ) seems relevant.

jakirkham commented:
Looks pretty good to me. Made a few comments above, but nothing essential. Would be nice to get rid of the `sys.path` hack if possible, but not crucial if there is some issue requiring it as a workaround.

alimanfoo commented:
Thanks for review, would be good to know what you do instead of sys.path hack.

jakirkham commented:
Normally I've been doing `pip install -e .` when working locally, which ensures the development version is on the path. If that's not working, would be curious to know more details about the issue that is causing the need for the hack.

alimanfoo commented:
Thanks for the tip re pip install -e, I've updated the demo notebook. Also added support for ``expand`` being an int, so can control to what depth tree is expanded. I think I'll leave the rest as-is, having some of these methods broken out as functions is convenient, especially where long lines are needed for embedded html/javascript.

jakirkham commented:
Thanks @alimanfoo. Sounds reasonable. We can always go back and tweak later if needed.
